components/calendar/test/calendar.visual-diff.js

```diff
@@ -0,0 +1,272 @@
+				await page.keyboard.press('ArrowDown');
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+
+			it('navigates to next month when right arrow pressed enough times', async function() {
+				await tabToDates();
+				for (let i = 0; i < 18; i++) {
+					await page.keyboard.press('ArrowRight');
+				}
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+		});
+
+		describe('other key navigation', () => {
+			// the selected day is the monday with first day of week sunday
+			it('to the end of the week when END key pressed', async function() {
+				await tabToDates();
+				await page.keyboard.press('End');
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+
+			it('to the start of the week when HOME key pressed', async function() {
+				await tabToDates();
+				await page.keyboard.press('Home');
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+
+			it('to first week next month when PAGEDOWN pressed from month with more weeks than next', async function() {
+				await page.$eval(firstCalendarOfPage, (calendar) => {
+					const arrow1 = calendar.shadowRoot.querySelector('d2l-button-icon[text="Show January"]');
+					arrow1.click();
+					const arrow2 = calendar.shadowRoot.querySelector('d2l-button-icon');
+					arrow2.click();
+				});
+				await tabToDates();
+				await page.keyboard.press('PageDown');
+
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+
+			it('to last week in previous month when PAGEUP pressed from month with more weeks than next', async function() {
+				await page.$eval(firstCalendarOfPage, (calendar) => {
+					const arrow1 = calendar.shadowRoot.querySelector('d2l-button-icon[text="Show January"]');
+					arrow1.click();
+					const arrow2 = calendar.shadowRoot.querySelector('d2l-button-icon');
+					arrow2.click();
+				});
+
+				await page.$eval(firstCalendarOfPage, (calendar) => {
+					const date = calendar.shadowRoot.querySelector('div[data-date="3"][data-month="0"]');
+					date.click();
+				});
+
+				await tabToDates();
+				await page.keyboard.press('PageUp');
+
+				const rect = await visualDiff.getRect(page, firstCalendarOfPage);
+				await visualDiff.screenshotAndCompare(page, this.test.fullTitle(), { clip: rect });
+			});
+
+		});
+	});
+
+	const tabToDates = async function() {
+		await page.keyboard.press('Tab');
+		await page.keyboard.press('Tab');
+		await page.keyboard.press('Tab');
+	};
```

Comment:
Normally I don't think it's super valuable to do visual diffs in every language, but in this case since there's so much rendering that depends on the locale settings I think it might be worth it? Not every permutation in every language, but just a standard calendar view in each one maybe?

Reply:
I'm struggling to get this to work. Have you seen an example of this working? Basically I just can't seem to be able to update the language and then have that be what's reflected on the visual-diff page (it still shows english). The only way that seems like it would currently work (unless I've missed something) is to have a new `.visual-diff.html` page for each `lang` which has the corresponding `lang` set on `html` and that doesn't seem ideal (though it there isn't a better way then that is fairly straightforward to do).

Reply:
I would expect something like this to work:

```
import { getDocumentLocaleSettings, supportedLocales } from '@brightspace-ui/intl/lib/common.js';

describe('Test Suite', () => {

	const documentLocaleSettings = getDocumentLocaleSettings();
	afterEach(() => documentLocaleSettings.reset();

	supportedLocales.forEach((lang) => {
		it(`should work for ${lang}`, () => {
			documentLocaleSettings.language = lang;
			await oneEvent(elem, 'd2l-localize-behavior-language-changed');
			// take visual diff
		});
	});

});
```

Not sure if you need the `await oneEvent(...)` or not.

Reply:
I don't think we can currently use `import` with `visual-diff` so I don't think we can use changing `documentLocaleSettings`. I have mostly tried different variants of `await page.evaluate(() =>document.querySelector('html').setAttribute('lang', 'ar'));` and some other options I have found online but haven't had luck so far.

Reply:
Ah, right. What about combining the `setAttribute` to change the language with the `await` for the `d2l-localize-behavior-language-changed` event?

Reply:
If you can get this working, this is probably something we could refactor into the visual-diff code to make it easier for others. @dbatiste FYI.

Reply:
Ooo I figured it out. Updating `lang` in this way does actually trigger all the right property updates etc. but because `calendarData` is already instantiated for `en`, it does not get updated. I can listen for that event in `calendar.js` and update `calendarData`

Reply:
Nice! 🎉